zrml/prediction-markets/src/lib.rs

```diff
@@ -2128,7 +2131,8 @@ mod pallet {
                 bonds.total_amount_bonded(&who),
             )?;
 
-            let market_id = <zrml_market_commons::Pallet<T>>::push_market(market.clone())?;
+            let (market_id, market) =
+                <zrml_market_commons::Pallet<T>>::build_market(market_builder)?;
```

Comment:
I would simply avoid the builder pattern by peeking the next market id, then construct the market structure and then push the market.

Reply:
Why don't we even decouple market id creation from `push_market`?

Reply:
So basically just throw `let market_id = Self::next_market_id()?;` out of `push_market` and use it directly.

Reply:
Again, builder pattern makes total sense, if you just care about a few fields for example in mocks, benchmarking and tests to get to a working structure quick and not having the full structure initialisation and associated maintenance if it changes all the time.

Reply:
I also generally seem to derive a sense of a will of you to put all sorts of "clean code" patterns in the business logic without taking a step back to reconsider if those patterns are really appropriate for this specific scenario.

Reply:
> Why don't we even decouple market id creation from push_market?

Because they belong together. Placing the `Market` object in storage is the responsibility of market-commons. Part of that is ensuring that 1) the correct next market ID is used; 2) (this is new) the storage key matches the `market_id` field of the market.

If we let other modules set the market ID, we take that responsiblity away from market-commons and place it in the hands of the callers. This risks that they put the wrong market ID into the market struct, overwrite old markets or skip market IDs. This is just basic encapsulation.

To maintain encapsulation, we need to be able to pass the market creation parameters to market-commons without setting the `market_id` field. The builder pattern seemed appropriate to me.

Moreover, as I've pointed out in the description, the builder pattern will make even more sense when we get rid of `edit_market` and can move valdiation functionality into the trait implementation. This allows us to move validation out of extrinsics and free functions of a bloated `lib.rs` into separate objects.

And once `edit_market` is done for, we can even remove certain setters. For example, why would we have a setter for the `early_close` field? Clearly, a newly created market should never have an early close scheduled, right? These will be removed as soon as `edit_market` is gone. This further encapsulates the market creation.

> I also generally seem to derive a sense of a will of you to put all sorts of "clean code" patterns in the business logic without taking a step back to reconsider if those patterns are really appropriate for this specific scenario and couldn't be done more simple.

I'm indeed very concerned with increasing the code quality of our code base. As for "[not] taking a step back", I'm surprised I'm making that impression. Not sure what to tell you except that I always weigh the available options very carefully.

Reply:
I understand your point about encapsulation. My approach wasn't meant to be the best either ;-), but a first attempt. Yeah, this wasn't in my mind, but I still don't like the `ImcompleteMarketBuilder` error either. And the fact that we have a "complex" market builder because of one field for this huge struct. Another option would be to have a function with a lot of parameters for all market fields or a `PartialMarket` struct passed to this function.

> Moreover, as I've pointed out in the description, the builder pattern will make even more sense when we get rid of edit_market and can move valdiation functionality into the trait implementation. This allows us to move validation out of extrinsics and free functions of a bloated lib.rs into separate objects.

This sounds interesting, I like the idea a lot. Why not using a function `check_market(market)` and validate everything inside it? Then we could also get rid of the bloated lib.rs. I mean the logic of validation needs to sit somewhere anyways. 

> And once edit_market is done for, we can even remove certain setters. For example, why would we have a setter for the early_close field? Clearly, a newly created market should never have an early close scheduled, right? These will be removed as soon as edit_market is gone. This further encapsulates the market creation.

This is a good point. However for market creation, you could just use a `PartialMarket` without an early close field and assign `None` to it inside the market-commons pallet. In that way the market's early close hasn't to be assigned in the prediction markets pallet either.

> As for "[not] taking a step back", I'm surprised I'm making that impression.

That's why I said `seem`. It's very subtle and I wanted to alert about things early to improve overall readability and wanted to challenge assumptions. At the end, try to make you better ;-) nothing else. It was never about me blaming you.

> Not sure what to tell you except that I always weigh the available options very carefully.

❤️

Reply:
The whole purpose of a builder pattern is to be capable to create different representations of the same object in a clean and structured manner. So my question is, is there any foreseen utilization of that feature? Will the `Market` object ever be created with different fields set or not set? If the answer is yes, then this is the way to go. If the answer is no, then simply create a constructor for `Market` that is invoked in `market-commons` within a function that resembles the constructor but does not contain the `market_id` (as it is derived within that function).

I am not quite sure how updating markets is planned to work in the future, but I think in any case it makes sense to apply a good design pattern and make the fields of `Market` private while providing getters for all fields and setters for fields that can change during the lifetime of a market.

Reply:
That's not the only purpose of the builder struct. Here, the purpose is to allow the `Market` object to be created piece-by-piece across multiple functions. This is the case here. It also allows placing the validation logic somewhere outside of the prediction-marktes pallet.

I think I've made my point here and in the long discussion above. Could somebody else please make a decision so we can get this out the door?

Reply:
I see, if that's the requirement than this is the correct design choice. Move forward.